pyrokinetics/gk_code/GKInputReaderGS2.py

```diff
@@ -0,0 +1,244 @@
+from ..numerics import Numerics
+from .GKInputReader import GKInputReader
+from .gs2_utils import pyro_gs2_miller, pyro_gs2_species, gs2_is_nonlinear
+
+
+class GKInputReaderGS2(GKInputReader):
+    """
+    Class that can read GS2 input files, and produce
+    Numerics, LocalSpecies, and LocalGeometry objects
+    """
+
+    code_name = "GS2"
+
+    def read(self, filename: PathLike) -> f90nml.Namelist:
+        """
+        Reads GS2 input file into a dictionary
+        """
+        super().read(filename)
+        # TODO These lines correct a possible mistake in the input file. Why is
+        #      this mistake corrected in particular? Should these corrections be
+        #      expanded?
+        if self.is_nonlinear() and "wstar_units" in self.data["knobs"]:
+            self.data["knobs"].pop("wstar_units")
+        return self.data
+
+    def verify(self, filename: PathLike):
+        """
+        Ensure this file is a valid gs2 input file"
+        """
+        data = f90nml.read(filename).todict()
+        expected_keys = ["kt_grids_knobs", "theta_grid_knobs", "theta_grid_eik_knobs"]
+        if not np.all(np.isin(expected_keys, list(data))):
+            raise ValueError(f"Expected GS2 file, received {filename}")
+
+    def is_nonlinear(self) -> bool:
+        return gs2_is_nonlinear(self.data)
+
+    def add_flags(self, flags) -> None:
+        """
+        Add extra flags to GS2 input file
+        """
+        for key, parameter in flags.items():
+            if key not in self.data:
+                self.data[key] = dict()
+            for param, val in parameter.items():
+                self.data[key][param] = val
+
+    def get_local_geometry(self) -> LocalGeometry:
+        """
+        Returns local geometry. Delegates to more specific functions
+        """
+        gs2_eq = self.data["theta_grid_knobs"]["equilibrium_option"]
+        if gs2_eq in ["eik", "default"]:
+            local_eq = self.data["theta_grid_eik_knobs"].get("local_eq", True)
+            iflux = self.data["theta_grid_eik_knobs"].get("iflux", 0)
+            if local_eq:
+                if iflux == 0:
+                    return self.get_local_geometry_miller()
+                else:
+                    # return self.get_fourier()
+                    raise NotImplementedError("GS2 Fourier options are not implemented")
+            else:
+                raise RuntimeError("GS2 is not using local equilibrium")
+        else:
+            raise NotImplementedError(
+                f"GS2 equilibrium option {gs2_eq} not implemented"
+            )
+
+    def get_local_geometry_miller(self) -> LocalGeometryMiller:
+        """
+        Load Miller object from GS2 file
+        """
+        # TODO can we get psi_n from this?
+
+        # Set some defaults here
+        # FIXME Why do we modify self as a side effect?
+        self.data["theta_grid_eik_knobs"]["bishop"] = 4
+        self.data["theta_grid_eik_knobs"]["irho"] = 2
```

Comment:
Ah, but this when reading the file. Should we instead error if these options are not what we expect/need? If the input file has these set to non-default values, are we changing the meaning of some of the other options?

I guess these should really be set in the writer?

Reply:
It shouldn't really matter what they are in the template. As long as they are set correctly when we write the file then we should be fine. But agreed that this could be moved to the writer.

Only caveat I guess is when reading in existing data that doesn't have this set. Then pyrokinetics can't deal with that output correctly